bdk_chain/src/chain_graph.rs

```diff
@@ -55,13 +56,27 @@ impl<P: ChainPosition> ChainGraph<P> {
     /// Inserts a transaction into the inner [`ChainGraph`] and optionally into the chain at
     /// `position`.
     ///
+    /// If inserting it into a chain `position` will result in a conflict, an error will be returned
+    /// instead.
+    ///
     /// **Warning**: This function modifies the internal state of the chain graph. You are
     /// responsible for persisting these changes to disk if you need to restore them.
     pub fn insert_tx(
         &mut self,
         tx: Transaction,
         position: Option<P>,
-    ) -> Result<bool, sparse_chain::InsertTxErr> {
+    ) -> Result<bool, InsertTxErr<P>> {
+        // only care about conflicts if chain position is provided
+        if position.is_some() {
+            let conflicts = self
+                .conflicting_txids(&tx)
+                .map(|(k, v)| (k.clone(), v))
+                .collect::<BTreeMap<_, _>>();
+            if !conflicts.is_empty() {
+                return Err(InsertTxErr::Conflicts(conflicts));
+            }
+        }
+
         let chain_changed = match position {
             Some(pos) => self.chain.insert_tx(tx.txid(), pos)?,
             None => false,
```

Comment:
To me it seems if `position` is `None`, we will add it only to graph and not in chain. Is that intended? And then we can add conflicting transactions too in the graph, by bypassing the filter above?